packages/@sanity/code-input/src/CodeInput.tsx

```diff
@@ -86,222 +102,194 @@ function isSupportedLanguage(mode: string) {
+          PatchEvent.from([setIfMissing({_type: type.name}), val ? set(val, path) : unset(path)])
+        )
+      },
+      [onChange, type.name]
     )
-  }
-
-  handleToggleSelectLine = (lineNumber: number) => {
-    const {type, onChange, value} = this.props
-    const path: Array<string | number> = ['highlightedLines']
-    const highlightedLines = (value && value.highlightedLines) || []
-
-    let position = highlightedLines.indexOf(lineNumber)
-    const patches = [setIfMissing({_type: type.name}), setIfMissing([], ['highlightedLines'])]
-    const addLine = position === -1
-
-    if (addLine) {
-      // New element, figure out where to add it so it sorts correctly
-      const sorted = highlightedLines.concat(lineNumber).sort(compareNumbers)
-      position = sorted.indexOf(lineNumber)
-      patches.push(
-        insert([lineNumber], 'before', path.concat(position === sorted.length - 1 ? -1 : position))
-      )
-    } else if (highlightedLines.length === 1) {
-      // Last element removed, unset whole path
-      patches.push(unset(path))
-
-      // Temporary workaround for bug in react-ace
-      // (https://github.com/securingsincity/react-ace/issues/229)
-      const editor = this.editor
-
-      // Remove all markers from editor
-      ;[true, false].forEach((inFront) => {
-        const editorSession = editor?.getSession()
-        const currentMarkers = editorSession.getMarkers(inFront)
-        Object.keys(currentMarkers).forEach((marker) => {
-          editorSession.removeMarker(currentMarkers[marker].id)
-        })
-      })
-    } else {
-      // Removed, but not the last element, remove single item
-      patches.push(unset(path.concat(position)))
-    }
-
-    onChange(PatchEvent.from(patches))
-  }
-
-  handleGutterMouseDown = (event: any) => {
-    const target = event.domEvent.target
-    if (target.classList.contains('ace_gutter-cell')) {
-      const row = event.getDocumentPosition().row
-      this.handleToggleSelectLine(row + 1) // Ace starts at row 0
-    }
-  }
 
-  handleEditorLoad = (editor: any) => {
-    this.editor = editor
-
-    editor.focus()
-    editor.on('guttermousedown', this.handleGutterMouseDown)
-  }
-
-  handleLanguageChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
-    const {type, onChange} = this.props
-    const value = event.currentTarget.value
-    const path = PATH_LANGUAGE
-
-    onChange(
-      PatchEvent.from([setIfMissing({_type: type.name}), value ? set(value, path) : unset(path)])
+    const getTheme = useCallback(() => {
+      const preferredTheme = get(type, 'options.theme')
+      return preferredTheme && SUPPORTED_THEMES.find((theme) => theme === preferredTheme)
+        ? preferredTheme
+        : DEFAULT_THEME
+    }, [type])
+
+    const handleToggleSelectLine = useCallback(
+      (lineNumber: number) => {
+        const editorSession = aceEditorRef.current?.editor?.getSession()
+        const backgroundMarkers = editorSession.getMarkers(true)
```

Comment:
`editorSession` may be undefined at this point (at least type-wise), so perhaps a good idea to add a guard here (note to self I guess, since I wrote this 😆).